src/architecture/unixfs.md

```diff
@@ -0,0 +1,618 @@
+remaining components and on the CID you popped.
+
+
+#### `Symlink` type
+
+A :dfn[Symlink] represents a POSIX [symbolic link](https://pubs.opengroup.org/onlinepubs/9699919799/functions/symlink.html).
+A symlink MUST NOT have children. <!--TODO: check that this is true-->
+
+The `PBNode.Data.Data` field is a POSIX path that MAY be appended in front of the
+currently remaining path component stack.
+
+##### Path Resolution
+
+There is no current consensus on how pathing over symlinks should behave. Some
+implementations return symlink objects and fail if a consumer tries to follow them
+through.
+
+Following the POSIX specification over the current UnixFS path context is probably fine.
+
+#### `HAMTDirectory`
+
+A :dfn[HAMT Directory] is a [Hashed-Array-Mapped-Trie](https://en.wikipedia.org/wiki/Hash_array_mapped_trie)
+data structure representing a :ref[Directory]. It is generally used to represent
+directories that cannot fit inside a single block. They are also known as "sharded
+directories" since they allow to split large directories into multiple blocks, the "shards".
+
+- `decode(PBNode.Data).hashType` indicates the [multihash] function to use to digest
+  the path components used for sharding. It MUST be `murmur3-x64-64` (`0x22`).
+- `decode(PBNode.Data).Data.Data` is a bit field, which indicates whether or not
+  links are part of this HAMT, or its leaves. The usage of this field is unknown given
+  that you can deduce the same information from the link names.
+- `decode(PBNode.Data).Data.fanout` MUST be a power of two. This encodes the number
+  of hash permutations that will be used on each resolution step. The log base 2
+  of the `fanout` indicate how wide the bitmask will be on the hash at for that step.
+  `fanout` MUST be between 8 and probably 65536. <!-- 65536 is a totally arbitrary choice I made, FIXME: get consensus on an upper bound. -->.
+
+The field `Name` of an element of `PBNode.Links` for a HAMT starts with an
+uppercase hex-encoded prefix, which is `log2(fanout)` bits wide.
+
+##### Path Resolution
+
+To resolve the path inside a HAMT:
+
+1. Take the current path component then hash it using the [multihash] represented
+   by the value of `decode(PBNode.Data).hashType`.
+2. Pop the `log2(fanout)` lowest bits from the path component hash digest, then
+   hex encode (using 0-F) those bits using little endian. Find the link that starts
+   with this hex encoded path.
+3. If the link `Name` is exactly as long as the hex encoded representation, follow
+   the link and repeat step 2 with the child node and the remaining bit stack.
+   The child node MUST be a HAMT directory else the directory is invalid, else continue.
+4. Compare the remaining part of the last name you found, if it match the original
+   name you were trying to resolve you successfully resolved a path component,
+   everything past the hex encoded prefix is the name of that element
+   (useful when listing children of this directory).
```

Comment:
```suggestion
3. If the link `Name` is exactly as long as the hex encoded representation, follow
   the link and repeat step 2 with the child node and the remaining bit stack.
   The child node MUST be a HAMT directory, or else the directory is invalid. Otherwise, continue.
4. Compare the remaining part of the last name you found. If it matches the original
   name you were trying to resolve, you have successfully resolved a path component.
   Everything past the hex encoded prefix is the name of that element, which is useful when listing children of this directory.
```